default/python/AI/ColonisationAI.py

```diff
@@ -1344,3 +1332,116 @@ def get_first_column_value(x):
             ])
     print
     candidate_table.print_table()
+
+
+class OrbitalBaseManager(object):
+    base_type = "INVALID"
+    role = ShipRoleType.INVALID
+
+    def __init__(self):
+        self.qualifying_targets = {}
+        self.target_candidates = {}
+        self.num_enqueued_bases = 0
+
+    def reset(self):
+        self.__init__()
+
+    def get_all_qualifying_targets(self):
+        return self.qualifying_targets.keys()
+
+    def log_as_qualifying_target(self, target, build_location):
+        if target not in self.qualifying_targets:
+            self.qualifying_targets[target] = [build_location, -1]
+
+    def mark_as_enqueued(self, target):
+        self.qualifying_targets[target][1] = self.qualifying_targets[target][0]
+
+    def is_enqueued(self, target):
+        return self.qualifying_targets[target][1] != -1
+
+    def get_building_loc(self, target):
+        return self.qualifying_targets[target][0]
+
+    def _enqueued_base_has_valid_target(self, build_location):
+        """Tries to assign a target planet for a base built at the specified location."""
+        # Try to find the target planet for which this base was enqueued
+        original_target = next((target for target, (loc, built) in self.target_candidates.items()
+                                if loc == build_location and built == build_location), None)
+
+        if original_target:
+            debug("%s base built at %d still has its original target." % (
+                self.base_type, build_location))
+            # do not assign the target of this base to another base
+            del self.target_candidates[original_target]
+            return True
+
+        # The original target seems to be no longer valid - this could happen
+        # if it was already claimed (possibly by another player).
+        # In this case, we try to find another valid target within the same
+        # which has not already been assigned a base.
+        alternative_target = next((target for target, (loc, built) in self.target_candidates.items()
+                                   if loc == build_location and built == -1), None)
+        if alternative_target:
+            debug("Reassigning %s base built at %d to new target %d as old target is no longer valid" % (
+                self.base_type, build_location, alternative_target))
+
+            # register the new target for the base so we do not build another
+            # base for the target.
+            self.qualifying_targets[alternative_target] = [build_location, build_location]
+
+            # do not assign the target of this base to another base.
+            del self.target_candidates[alternative_target]
+            return True
+
+        # could not find a target for the base
+        return False
+
+    def handle_bases_in_production_queue(self):
```

Comment:
Looks like it is part of initializer so lets make it private and move to `__init__`.